File: android/app/src/main/java/com/noop/analytics/FormulaPublicationGate.kt
package com.noop.analytics

import android.content.SharedPreferences

/** Upgrade boundary for Charge formula changes that do not create a raw-input dirty marker. */
internal object ChargeFormulaUpgradeGate {
    const val COMPLETED_REVISION_KEY = "noop.analysis.completedChargeFormulaRevision"
    const val HISTORY_DAYS = 4_000
    const val CURRENT_REVISION = NoopScoreAlgorithmRevision.CHARGE

    fun needsRescore(completedRevision: String?): Boolean =
        completedRevision != CURRENT_REVISION
}

/**
 * Upgrade boundary for Rest formula changes that do not create a raw-input dirty marker.
 *
 * Completion is written only after every resolvable historical day and required pass boundary
 * succeeds.
 */
internal object RestFormulaUpgradeGate {
    const val COMPLETED_REVISION_KEY = "noop.analysis.completedRestFormulaRevision"
    const val NEXT_ANCHOR_KEY = "noop.analysis.restFormulaNextAnchor"
    const val ANCHOR_REVISION_KEY = "noop.analysis.restFormulaAnchorRevision"
    const val HISTORY_DAYS = 4_000
    const val CURRENT_REVISION = NoopScoreAlgorithmRevision.REST
    val TRAVERSAL_REVISION =
        "${ChargeFormulaUpgradeGate.CURRENT_REVISION}|$CURRENT_REVISION"

    sealed interface Progress {
        data object Retry : Progress
        data class Advance(val nextAnchor: Long) : Progress
        data class Complete(val revision: String) : Progress
    }

    fun needsRescore(completedRevision: String?): Boolean =
        completedRevision != CURRENT_REVISION

    fun traversalAnchor(
        migrationRequired: Boolean,
        anchorRevision: String?,
        storedAnchor: Long?,
    ): Long? =
        storedAnchor?.takeIf {
            migrationRequired &&
                (
                    anchorRevision == TRAVERSAL_REVISION ||
                        anchorRevision == CURRENT_REVISION
                    ) &&
                it >= 0L
        }

    fun progress(
        passCompleted: Boolean,
        resolvableHistorySatisfied: Boolean,
        nextResolvableHistoryAnchor: Long?,
        wasRequired: Boolean,
        traversalWasSelected: Boolean,
    ): Progress {
        if (!passCompleted || !wasRequired || !traversalWasSelected) {
            return Progress.Retry
        }
        if (resolvableHistorySatisfied && nextResolvableHistoryAnchor == null) {
            return Progress.Complete(CURRENT_REVISION)
        }
        return nextResolvableHistoryAnchor
            ?.takeIf { it >= 0L }
            ?.let(Progress::Advance)
            ?: Progress.Retry
    }
}

/**
 * Fail-closed publication policy for locally computed derived metrics during formula migration.
 *
 * Raw, imported, journal, and official-reference namespaces remain publishable. Only the
 * `noop_computed` derived namespace waits until both current completion markers are durable.
 */
internal object FormulaPublicationGate {
    const val COMPUTED_SOURCE_KIND = "noop_computed"
    const val MANAGED_DERIVED_DATA_CLASS = "derived_summaries"
    val DEFERRED_DIAGNOSTIC_FIELDS = mapOf(
        "computed_derived" to "deferred",
        "reason" to "formula_migration",
    )

    fun computedDerivedReady(preferences: SharedPreferences): Boolean =
        computedDerivedReady(
            completedChargeRevision = preferences.getString(
                ChargeFormulaUpgradeGate.COMPLETED_REVISION_KEY,
                null,
            ),
            completedRestRevision = preferences.getString(
                RestFormulaUpgradeGate.COMPLETED_REVISION_KEY,
                null,
            ),
        )

    fun computedDerivedReady(
        completedChargeRevision: String?,
        completedRestRevision: String?,
    ): Boolean =
        completedChargeRevision == ChargeFormulaUpgradeGate.CURRENT_REVISION &&
            completedRestRevision == RestFormulaUpgradeGate.CURRENT_REVISION

    fun shouldPublishDerived(
        sourceKind: String,
        computedDerivedReady: Boolean,
    ): Boolean =
        sourceKind != COMPUTED_SOURCE_KIND || computedDerivedReady

    fun managedDataClasses(
        sourceKind: String,
        available: List<String>,
        computedDerivedReady: Boolean,
    ): List<String> =
        if (shouldPublishDerived(sourceKind, computedDerivedReady)) {
            available
        } else {
            available.filterNot { it == MANAGED_DERIVED_DATA_CLASS }
        }
}

File: Strand/System/FormulaPublicationGate.swift
import Foundation
import StrandAnalytics

/// Upgrade boundary for formula changes that do not alter raw-input fingerprints.
///
/// A revision string, rather than a one-shot boolean, makes every future Charge revision fail open
/// into a full-history rescore. Completion is persisted only after the analysis pass succeeds.
enum ChargeFormulaUpgradeGate {
    static let completedRevisionKey = "noop.analysis.completedChargeFormulaRevision"
    static let historyDays = 4_000
    static var currentRevision: String { NoopScoreAlgorithmRevision.charge }

    static func needsRescore(completedRevision: String?) -> Bool {
        completedRevision != currentRevision
    }

    static func revisionToPersist(passCompleted: Bool, wasRequired: Bool) -> String? {
        passCompleted && wasRequired ? currentRevision : nil
    }
}

/// Upgrade boundary for Rest formula changes that do not alter raw-input fingerprints.
///
/// Completion is persisted only after every resolvable historical day and required pass boundary
/// succeeds, so old and new Sleep Scores never share one current-revision publication label.
enum RestFormulaUpgradeGate {
    static let completedRevisionKey = "noop.analysis.completedRestFormulaRevision"
    static let nextAnchorKey = "noop.analysis.restFormulaNextAnchor"
    static let anchorRevisionKey = "noop.analysis.restFormulaAnchorRevision"
    static let historyDays = 4_000
    static var currentRevision: String { NoopScoreAlgorithmRevision.rest }
    static var traversalRevision: String {
        "\(ChargeFormulaUpgradeGate.currentRevision)|\(currentRevision)"
    }

    enum Progress: Equatable {
        case retry
        case advance(nextAnchor: Int)
        case complete(revision: String)
    }

    static func needsRescore(completedRevision: String?) -> Bool {
        completedRevision != currentRevision
    }

    static func traversalAnchor(
        migrationRequired: Bool,
        anchorRevision: String?,
        storedAnchor: Int?
    ) -> Int? {
        guard migrationRequired,
              anchorRevision == traversalRevision
                || anchorRevision == currentRevision,
              let storedAnchor,
              storedAnchor >= 0 else {
            return nil
        }
        return storedAnchor
    }

    static func progress(
        receipt: IntelligenceEngine.ScoreRunReceipt?,
        wasRequired: Bool,
        traversalWasSelected: Bool
    ) -> Progress {
        guard wasRequired,
              traversalWasSelected,
              let receipt,
              receipt.allRequiredBoundariesCompleted else {
            return .retry
        }
        if receipt.resolvableHistorySatisfied,
           receipt.nextResolvableHistoryAnchor == nil {
            return .complete(revision: currentRevision)
        }
        if let nextAnchor = receipt.nextResolvableHistoryAnchor,
           nextAnchor >= 0 {
            return .advance(nextAnchor: nextAnchor)
        }
        return .retry
    }
}

/// Fail-closed publication policy for locally computed derived metrics during formula migration.
///
/// Raw, imported, journal, and official-reference namespaces remain publishable. Only the
/// `noop_computed` derived namespace waits until both current migration markers are durable.
enum FormulaPublicationGate {
    static let computedSourceKind = "noop_computed"
    static let managedDerivedDataClass = "derived_summaries"
    static let deferredDiagnosticFields = [
        "computed_derived": "deferred",
        "reason": "formula_migration",
    ]

    static func computedDerivedReady(defaults: UserDefaults = .standard) -> Bool {
        computedDerivedReady(
            completedChargeRevision: defaults.string(
                forKey: ChargeFormulaUpgradeGate.completedRevisionKey
            ),
            completedRestRevision: defaults.string(
                forKey: RestFormulaUpgradeGate.completedRevisionKey
            )
        )
    }

    static func computedDerivedReady(
        completedChargeRevision: String?,
        completedRestRevision: String?
    ) -> Bool {
        completedChargeRevision == ChargeFormulaUpgradeGate.currentRevision
            && completedRestRevision == RestFormulaUpgradeGate.currentRevision
    }

    static func shouldPublishDerived(
        sourceKind: String,
        computedDerivedReady: Bool
    ) -> Bool {
        sourceKind != computedSourceKind || computedDerivedReady
    }

    static func shouldPublishSocialSummaries(
        computedDerivedReady: Bool
    ) -> Bool {
        computedDerivedReady
    }

    static func publishSocialSummariesIfReady<T>(
        computedDerivedReady: Bool,
        publish: () async throws -> T
    ) async rethrows -> T? {
        guard shouldPublishSocialSummaries(
            computedDerivedReady: computedDerivedReady
        ) else {
            return nil
        }
        return try await publish()
    }

    static func managedDataClasses(
        sourceKind: String,
        available: [String],
        computedDerivedReady: Bool
    ) -> [String] {
        guard !shouldPublishDerived(
            sourceKind: sourceKind,
            computedDerivedReady: computedDerivedReady
        ) else {
            return available
        }
        return available.filter { $0 != managedDerivedDataClass }
    }
}
